Description: ugly hack to add a UEFI firmware option
 This patch is an ugly hack to add an option in virt-manager for
 installing a VM with the OVMF UEFI firmware. Ideally, for this to be
 done in a clean way:
 1- qemu needs to be able to report what firmware it locates and supports
 2- the libvirt qemu driver needs to report that as capabilities
 3- virt-manager needs to probe for firmware via capabilities, and display
    a dynamic list
Author: Marc Deslauriers <marc.deslauriers@canonical.com>
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/virt-manager/+bug/1074207
Forwarded: no, not ready
Updated: 2014-02-17

Index: virt-manager-1.1.0/ui/create.ui
===================================================================
--- virt-manager-1.1.0.orig/ui/create.ui	2014-09-09 23:09:37.956571415 +0200
+++ virt-manager-1.1.0/ui/create.ui	2014-09-09 23:09:37.948571415 +0200
@@ -491,6 +491,34 @@
                                       </packing>
                                     </child>
                                     <child>
+                                      <object class="GtkComboBox" id="config-loader">
+                                        <property name="visible">True</property>
+                                        <property name="can_focus">False</property>
+                                      </object>
+                                      <packing>
+                                        <property name="left_attach">1</property>
+                                        <property name="top_attach">3</property>
+                                        <property name="width">1</property>
+                                        <property name="height">1</property>
+                                      </packing>
+                                    </child>
+                                    <child>
+                                      <object class="GtkLabel" id="label149">
+                                        <property name="visible">True</property>
+                                        <property name="can_focus">False</property>
+                                        <property name="xalign">1</property>
+                                        <property name="label" translatable="yes">_Firmware:</property>
+                                        <property name="use_underline">True</property>
+                                        <property name="mnemonic_widget">config-loader</property>
+                                      </object>
+                                      <packing>
+                                        <property name="left_attach">0</property>
+                                        <property name="top_attach">3</property>
+                                        <property name="width">1</property>
+                                        <property name="height">1</property>
+                                      </packing>
+                                    </child>
+                                    <child>
                                       <object class="GtkComboBox" id="config-machine">
                                         <property name="visible">True</property>
                                         <property name="can_focus">False</property>
Index: virt-manager-1.1.0/ui/details.ui
===================================================================
--- virt-manager-1.1.0.orig/ui/details.ui	2014-09-09 23:09:37.956571415 +0200
+++ virt-manager-1.1.0/ui/details.ui	2014-09-09 23:18:07.780556249 +0200
@@ -954,6 +954,35 @@
                                           </packing>
                                         </child>
                                         <child>
+                                          <object class="GtkLabel" id="overview-loader">
+                                            <property name="visible">True</property>
+                                            <property name="can_focus">False</property>
+                                            <property name="xalign">0</property>
+                                            <property name="label">foo</property>
+                                            <property name="selectable">True</property>
+                                          </object>
+                                          <packing>
+                                            <property name="left_attach">1</property>
+                                            <property name="top_attach">3</property>
+                                            <property name="width">1</property>
+                                            <property name="height">1</property>
+                                          </packing>
+                                        </child>
+                                        <child>
+                                          <object class="GtkLabel" id="label188">
+                                            <property name="visible">True</property>
+                                            <property name="can_focus">False</property>
+                                            <property name="xalign">1</property>
+                                            <property name="label" translatable="yes">Firmware:</property>
+                                          </object>
+                                          <packing>
+                                            <property name="left_attach">0</property>
+                                            <property name="top_attach">3</property>
+                                            <property name="width">1</property>
+                                            <property name="height">1</property>
+                                          </packing>
+                                        </child>
+                                        <child>
                                           <object class="GtkLabel" id="overview-machine-label">
                                             <property name="visible">True</property>
                                             <property name="can_focus">False</property>
@@ -963,7 +992,7 @@
                                           </object>
                                           <packing>
                                             <property name="left_attach">0</property>
-                                            <property name="top_attach">3</property>
+                                            <property name="top_attach">4</property>
                                           </packing>
                                         </child>
                                         <child>
@@ -975,7 +1004,7 @@
                                           </object>
                                           <packing>
                                             <property name="left_attach">1</property>
-                                            <property name="top_attach">3</property>
+                                            <property name="top_attach">4</property>
                                           </packing>
                                         </child>
                                         <child>
Index: virt-manager-1.1.0/virtManager/create.py
===================================================================
--- virt-manager-1.1.0.orig/virtManager/create.py	2014-09-09 23:09:37.956571415 +0200
+++ virt-manager-1.1.0/virtManager/create.py	2014-09-09 23:18:59.860554700 +0200
@@ -311,6 +311,11 @@
         archList.set_row_separator_func(
             lambda m, i, ignore: m[i][0] is None, None)
 
+        loaderList = self.widget("config-loader")
+        loaderModel = Gtk.ListStore(str)
+        loaderList.set_model(loaderModel)
+        uiutil.set_combo_text_column(loaderList, 0)
+
         hyperList = self.widget("config-hv")
         hyperModel = Gtk.ListStore(str, str)
         hyperList.set_model(hyperModel)
@@ -516,6 +521,7 @@
         self.conn.invalidate_caps()
         self.change_caps()
         self.populate_hv()
+        self.populate_loader()
         self.populate_arch()
 
         show_arch = (self.widget("config-hv").get_visible() or
@@ -746,6 +752,16 @@
         else:
             lst.emit("changed")
 
+    def populate_loader(self):
+        loader_list = self.widget("config-loader")
+        model = loader_list.get_model()
+        model.clear()
+
+        model.append(['Default'])
+        model.append(['UEFI'])
+        default = 0
+        loader_list.set_active(default)
+
     def populate_conn_list(self, urihint=None):
         conn_list = self.widget("create-conn")
         model = conn_list.get_model()
@@ -953,6 +969,21 @@
     def get_config_name(self):
         return self.widget("create-vm-name").get_text()
 
+    def get_config_loader(self):
+        loader_list = self.widget("config-loader")
+        model = loader_list.get_model()
+
+        idx = loader_list.get_active()
+        if idx < 0:
+            return None
+
+        loader = model[idx][0]
+
+        if loader == 'UEFI':
+            return 'OVMF.fd'
+        else:
+            return None
+
     def get_config_machine(self):
         return uiutil.get_list_selection(self.widget("config-machine"), 0,
             check_visible=True)
@@ -1724,6 +1755,8 @@
                 if not self.validate_storage_page():
                     return False
 
+        self.guest.os.loader = self.get_config_loader()
+
         macaddr = None
         if self.widget("config-macaddr").get_sensitive():
             macaddr = self.widget("config-macaddr").get_text().strip()
Index: virt-manager-1.1.0/virtManager/details.py
===================================================================
--- virt-manager-1.1.0.orig/virtManager/details.py	2014-09-09 23:09:37.956571415 +0200
+++ virt-manager-1.1.0/virtManager/details.py	2014-09-09 23:09:37.956571415 +0200
@@ -2443,8 +2443,12 @@
         self.widget("overview-hv").set_text(self.vm.get_pretty_hv_type())
         arch = self.vm.get_arch() or _("Unknown")
         emu = self.vm.get_emulator() or _("None")
+        loader = self.vm.get_loader() or _("Default")
+        if loader == "OVMF.fd":
+            loader = "UEFI"
         self.widget("overview-arch").set_text(arch)
         self.widget("overview-emulator").set_text(emu)
+        self.widget("overview-loader").set_text(loader)
 
         # Machine settings
         machtype = self.vm.get_machtype()
Index: virt-manager-1.1.0/virtManager/domain.py
===================================================================
--- virt-manager-1.1.0.orig/virtManager/domain.py	2014-09-09 23:09:37.956571415 +0200
+++ virt-manager-1.1.0/virtManager/domain.py	2014-09-09 23:09:37.956571415 +0200
@@ -1081,6 +1081,8 @@
 
     def get_emulator(self):
         return self.get_xmlobj().emulator
+    def get_loader(self):
+        return self.get_xmlobj().os.loader
     def get_machtype(self):
         return self.get_xmlobj().os.machine
     def get_idmap(self):
